Skip & Log incompatible loadouts

Raffson opened the pull request:
A (hopefully temporary) fix #2320 for skipping incompatible loadouts.

I still think the better approach would be to handle this in pydcs. Writing a scanner for weapons/pylons data probably solves this issue for as long as the directory structure stays the same. The upside would be that Liberation doesn't need to make a new release (together with a new export for pydcs) whenever DCS updates.

Khopa commented on game/ato/loadouts.py:
```diff
@@ -189,6 +202,11 @@ def default_for_task_and_aircraft(
             dcs_unit_type.load_payloads()
             payload = dcs_unit_type.loadout_by_name(name)
             if payload is not None:
+                if not cls.valid_payload(payload):
```
Hello Raffson, 

I've got an issue with this, this exception is triggered when I click on Begin campaign :

```
Traceback (most recent call last):
  File "D:\Work\dcs_liberation\qt_ui\widgets\QTopPanel.py", line 150, in passTurn
    self.game.pass_turn(no_action=True)
  File "D:\Work\dcs_liberation\game\game.py", line 325, in pass_turn
    self.initialize_turn(events)
  File "D:\Work\dcs_liberation\game\game.py", line 408, in initialize_turn
    self.red.initialize_turn()
  File "D:\Work\dcs_liberation\game\coalition.py", line 186, in initialize_turn
    self.plan_missions()
  File "D:\Work\dcs_liberation\game\coalition.py", line 207, in plan_missions
    TheaterCommander(self.game, self.player).plan_missions(tracer)
  File "D:\Work\dcs_liberation\game\commander\theatercommander.py", line 83, in plan_missions
    result = self.plan(state)
  File "D:\Work\dcs_liberation\game\htn.py", line 83, in plan
    if task.preconditions_met(planning_state.state):
  File "D:\Work\dcs_liberation\game\commander\tasks\primitive\barcap.py", line 18, in preconditions_met
    return super().preconditions_met(state)
  File "D:\Work\dcs_liberation\game\commander\tasks\packageplanningtask.py", line 50, in preconditions_met
    return self.fulfill_mission(state)
  File "D:\Work\dcs_liberation\game\commander\tasks\packageplanningtask.py", line 102, in fulfill_mission
    self.package = fulfiller.plan_mission(
  File "D:\Work\dcs_liberation\game\commander\packagefulfiller.py", line 167, in plan_mission
    self.plan_flight(
  File "D:\Work\dcs_liberation\game\commander\packagefulfiller.py", line 85, in plan_flight
    if not builder.plan_flight(flight):
  File "D:\Work\dcs_liberation\game\commander\packagebuilder.py", line 57, in plan_flight
    flight = Flight(
  File "D:\Work\dcs_liberation\game\ato\flight.py", line 64, in __init__
    self.loadout = Loadout.default_for(self)
  File "D:\Work\dcs_liberation\game\ato\loadouts.py", line 189, in default_for
    return cls.default_for_task_and_aircraft(
  File "D:\Work\dcs_liberation\game\ato\loadouts.py", line 205, in default_for_task_and_aircraft
    if not cls.valid_payload(payload):
  File "D:\Work\dcs_liberation\game\ato\loadouts.py", line 138, in valid_payload
    pylons = payload["pylons"]
TypeError: list indices must be integers or slices, not str
```

To reproduce :
- start a new campaign on Battle for Golan Height map
- keep campaign default settings and defaults squadrons
- buy one of each available planes at Ramat David
 - click on begin campaign

I'm surprised mypy didn't catch this, the call to `cls.valid_payload` does seems invalid.

I think you might have missed this, since the call to `default_for_task_and_aircraft` is cached, and doesn't happen when you load an existing save iirc.

Regards, 

Khopa

Raffson replied:
I'll have a look at it right away since I'm waiting for pydcs to update again, hopefully a last time now 😂

Raffson replied:
@Khopa Ok this is really weird indeed, typing suggests I should be getting a dictionary and not a list. Not sure what's going on here 🤔 

EDIT: Nvm, I think I found it! 👍 
Not sure what mypy is doing here but I definitely made a mistake in `default_for_task_and_aircraft`

Raffson replied:
Should be resolved...

Raffson committed:
Skip & Log incompatible loadouts

Solution exists in using the 'get' method
of the "Weapons dictionary", and
subsequentially guarding against
None. Aside from that I created
a method to validate a payload,
which uses this None value to
determine validity.

DanAlbert commented on game/data/weapons.py:
```diff
@@ -53,8 +53,9 @@ def name(self) -> str:
         updated = Weapon.with_clsid(state["clsid"])
-        state.update(updated.__dict__)
-        self.__dict__.update(state)
+        if updated is not None:
```
This needs a comment. I'm not really sure what the goal is.

Raffson replied:
The goal is to prevent `updated.__dict__` from generating an exception if `updated is None`. Remember, `Weapon.with_clsid` can now return `None` which is used to identify incompatible (for whatever reason) weapons...

DanAlbert commented on game/data/weapons.py:
```diff
@@ -267,7 +268,9 @@ def for_aircraft(cls, aircraft: AircraftType, number: int) -> Pylon:
                     continue
-                allowed.add(Weapon.with_clsid(weapon["clsid"]))
+                allowed_weapon = Weapon.with_clsid(weapon["clsid"])
+                if allowed_weapon is not None:
```
Do we actually expect this? If we can't find a weapon for a pylon that DCS says is there, that sounds like a bug.

Raffson replied:
That's not necessarily DCS telling us it's there but rather pydcs telling us it doesn't know about the weapon and thus won't be able to load the weapon. I can think of two different cases:

- An old loadout that was scanned by pydcs with an invalid CLSID, invalid as in outdated/obsolete and is thus no longer present in the weapons map
- A new loadout that was scanned after a newly released version of DCS while not having updated pydcs yet, which was recently the case with TALDs

Also, may I remind you that it was you who suggested to ignore payloads that can't be loaded in #2320, so your comment is really confusing me...
